### ISSUE TITLE
Detect when asynchronous requests have finished.

### ISSUE THREAD

**Issue by HeyItsMario** [igrishaev/etaoin#201]
**Disclaimer**: _This may not be an issue and I may be over complicating things._

1. Using Etaoin I navigate to a SPA.
(The way the SPA is set up is as follows: It has a list of questions where the inputs can be a dropdown (select tag), an input field (input type text) and perhaps a checkbox. As I answer each question the browser makes XHR requests to get additional questions and adds them to the page).
2. I query the elements, perform a click/fill and then wait a predefined amount of time before moving on to the next question.

**Is there a way I can detect if the browser has finished making an ajax request?** At that point, I would know that the browser has the response needed to add the elements onto the page, which signals that I can start querying the elements I need.

Currently what I am doing to achieve this behavior is running.
`(js-execute driver js)` where `js` is javascript that adds a global flag to the window, such as `window.xhr-complete? = false` and setting the flag to true whenever an XMLHttpRequest has returned, (status == 4). 
Basically used [this](https://stackoverflow.com/a/20714771/2977742) to implement my solution. And then I call `(wait-predicate driver xhr-complete?)` where `xhr-complete?` is a function the looks at the flag, after each interaction (click/fill).

I don't really like this solution and it feels hacky to me. Perhaps there is a better/cleaner way to handle this with Etaoin that I am not aware of.

**Comment by igrishaev** [comment]
Started working on the issue, some bits to keep them in mind:

```
(def c (chrome {:capabilities {:loggingPrefs {:performance "all"} 
                                           :chromeOptions {:perfLoggingPrefs {:enableNetwork true
                                                               :enablePage true
                                                               :traceCategories "browser,devtools.timeline,devtools"
                                                               :bufferUsageReportingInterval 100}}}}))

(get-log-types c)

(defmethods get-logs
  [:chrome :phantom]
  [driver]
  (->> (execute {:driver driver
                 :method :post
                 :path [:session (:session @driver) :log]
                 :data {:type :performance}})
       :value
       (mapv process-log)))

```

**Comment by igrishaev** [comment]
the idea is, you can query network logs and check AJAX state, but it's not for sure yet. I'll find out how to cover that with API

**Comment by igrishaev** [comment]
```
(clojure.pprint/pprint (last (filter (fn [log] (-> log :message :method (= "Network.responseReceived"))) logs)))
{:message
 {:method "Network.responseReceived",
  :params
  {:frameId "71E8F527E41868DD41BCBF70AEFD7112",
   :loaderId "5790DDCB17D1ABDD2BEEA35F81CBEF31",
   :requestId "1000037133.22",
   :response
   {:connectionId 51,
    :remotePort 443,
    :protocol "h2",
    :fromDiskCache false,
    :fromServiceWorker false,
    :connectionReused true,
    :headers #,
    :remoteIPAddress "74.125.131.104",
    :securityState "secure",
    :securityDetails #,
    :status 204,
    :timing #,
    :statusText "",
    :mimeType "text/html",
    :url
    "https://www.google.com/gen_204?atyp=i&ct=&cad=udla=1&ei=gPkZXZSMG76HwPAPiPaw4AM&di=1160&lt=42058&ln=153924&zx=1561983364848",
    :encodedDataLength 45},
   :timestamp 1232704.886122,
   :type "Other"}},
 :webview "71E8F527E41868DD41BCBF70AEFD7112"}
```

**Comment by igrishaev** [comment]
```
#{"Network.loadingFinished" "Page.loadEventFired" "Network.requestWillBeSent" "Page.frameStoppedLoading" "Page.domContentEventFired" "Page.frameNavigated" "Network.responseReceived" "Page.frameStartedLoading" "Page.frameResized" "Network.dataReceived" "Network.resourceChangedPriority"}
```

**Comment by igrishaev** [comment]
```
[{:level ::info,
  :message
  {:method "Network.requestWillBeSent",
   :params
   {:wallTime 1.56198853664202E9,
    :hasUserGesture false,
    :request
    {:hasPostData true,
     :headers
     {:Content-Type "text/plain;charset=UTF-8",
      :Origin "https://www.google.com",
      :Referer "https://www.google.com/",
      :User-Agent
      "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36"},
     :initialPriority "VeryLow",
     :method "POST",
     :mixedContentType "none",
     :referrerPolicy "origin",
     :url
     "https://www.google.com/gen_204?atyp=csi&ei=gPkZXZSMG76HwPAPiPaw4AM&s=jsa&jsi=s,et.focus,n.iDPoPb,cn.5&zx=1561988536641"},
    :frameId "71E8F527E41868DD41BCBF70AEFD7112",
    :initiator
    {:stack
     {:callFrames
      [{:columnNumber 120,
        :functionName "s_zha.log",
        :lineNumber 582,
        :scriptId "25",
        :url
        "https://www.google.com/xjs/_/js/k=xjs.s.ru.zL4ese5rBGM.O/ck=xjs.s.-u89cqk529a2f.L.W.O/m=Fkg7bd,HcFEGb,IvlUe,MC8mtf,OF7gzc,RMhBfe,T4BAC,TJw5qb,TbaHGc,Y33vzc,cdos,hsm,iDPoPb,jsa,mvYTse,tg8oTe,uz938c,vWNDde,ws9Tlc,yQ43ff,d,csi/am=AAAAi4CdbgJB_r8CANgFACBwF9jIAhEIyFA1AAE/d=1/dg=2/br=1/rs=ACT90oHLt6EenBVAvKo6i3GbKovDrprVVw"}
       {:columnNumber 318,
        :functionName "s_Ke.g9b",
        :lineNumber 329,
        :scriptId "25",
        :url
        "https://www.google.com/xjs/_/js/k=xjs.s.ru.zL4ese5rBGM.O/ck=xjs.s.-u89cqk529a2f.L.W.O/m=Fkg7bd,HcFEGb,IvlUe,MC8mtf,OF7gzc,RMhBfe,T4BAC,TJw5qb,TbaHGc,Y33vzc,cdos,hsm,iDPoPb,jsa,mvYTse,tg8oTe,uz938c,vWNDde,ws9Tlc,yQ43ff,d,csi/am=AAAAi4CdbgJB_r8CANgFACBwF9jIAhEIyFA1AAE/d=1/dg=2/br=1/rs=ACT90oHLt6EenBVAvKo6i3GbKovDrprVVw"}
       {:columnNumber 235,
        :functionName "s_Ne.log",
        :lineNumber 337,
        :scriptId "25",
        :url
        "https://www.google.com/xjs/_/js/k=xjs.s.ru.zL4ese5rBGM.O/ck=xjs.s.-u89cqk529a2f.L.W.O/m=Fkg7bd,HcFEGb,IvlUe,MC8mtf,OF7gzc,RMhBfe,T4BAC,TJw5qb,TbaHGc,Y33vzc,cdos,hsm,iDPoPb,jsa,mvYTse,tg8oTe,uz938c,vWNDde,ws9Tlc,yQ43ff,d,csi/am=AAAAi4CdbgJB_r8CANgFACBwF9jIAhEIyFA1AAE/d=1/dg=2/br=1/rs=ACT90oHLt6EenBVAvKo6i3GbKovDrprVVw"}
       {:columnNumber 346,
        :functionName "s_Bg.log",
        :lineNumber 473,
        :scriptId "25",
        :url
        "https://www.google.com/xjs/_/js/k=xjs.s.ru.zL4ese5rBGM.O/ck=xjs.s.-u89cqk529a2f.L.W.O/m=Fkg7bd,HcFEGb,IvlUe,MC8mtf,OF7gzc,RMhBfe,T4BAC,TJw5qb,TbaHGc,Y33vzc,cdos,hsm,iDPoPb,jsa,mvYTse,tg8oTe,uz938c,vWNDde,ws9Tlc,yQ43ff,d,csi/am=AAAAi4CdbgJB_r8CANgFACBwF9jIAhEIyFA1AAE/d=1/dg=2/br=1/rs=ACT90oHLt6EenBVAvKo6i3GbKovDrprVVw"}
       {:columnNumber 68,
        :functionName "s_rja.Ca",
        :lineNumber 476,
        :scriptId "25",
        :url
        "https://www.google.com/xjs/_/js/k=xjs.s.ru.zL4ese5rBGM.O/ck=xjs.s.-u89cqk529a2f.L.W.O/m=Fkg7bd,HcFEGb,IvlUe,MC8mtf,OF7gzc,RMhBfe,T4BAC,TJw5qb,TbaHGc,Y33vzc,cdos,hsm,iDPoPb,jsa,mvYTse,tg8oTe,uz938c,vWNDde,ws9Tlc,yQ43ff,d,csi/am=AAAAi4CdbgJB_r8CANgFACBwF9jIAhEIyFA1AAE/d=1/dg=2/br=1/rs=ACT90oHLt6EenBVAvKo6i3GbKovDrprVVw"}
       {:columnNumber 425,
        :functionName "",
        :lineNumber 461,
        :scriptId "25",
        :url
        "https://www.google.com/xjs/_/js/k=xjs.s.ru.zL4ese5rBGM.O/ck=xjs.s.-u89cqk529a2f.L.W.O/m=Fkg7bd,HcFEGb,IvlUe,MC8mtf,OF7gzc,RMhBfe,T4BAC,TJw5qb,TbaHGc,Y33vzc,cdos,hsm,iDPoPb,jsa,mvYTse,tg8oTe,uz938c,vWNDde,ws9Tlc,yQ43ff,d,csi/am=AAAAi4CdbgJB_r8CANgFACBwF9jIAhEIyFA1AAE/d=1/dg=2/br=1/rs=ACT90oHLt6EenBVAvKo6i3GbKovDrprVVw"}
       {:columnNumber 328,
        :functionName "s_Vfa",
        :lineNumber 376,
        :scriptId "25",
        :url
        "https://www.google.com/xjs/_/js/k=xjs.s.ru.zL4ese5rBGM.O/ck=xjs.s.-u89cqk529a2f.L.W.O/m=Fkg7bd,HcFEGb,IvlUe,MC8mtf,OF7gzc,RMhBfe,T4BAC,TJw5qb,TbaHGc,Y33vzc,cdos,hsm,iDPoPb,jsa,mvYTse,tg8oTe,uz938c,vWNDde,ws9Tlc,yQ43ff,d,csi/am=AAAAi4CdbgJB_r8CANgFACBwF9jIAhEIyFA1AAE/d=1/dg=2/br=1/rs=ACT90oHLt6EenBVAvKo6i3GbKovDrprVVw"}
       {:columnNumber 238,
        :functionName "s_ff",
        :lineNumber 374,
        :scriptId "25",
        :url
        "https://www.google.com/xjs/_/js/k=xjs.s.ru.zL4ese5rBGM.O/ck=xjs.s.-u89cqk529a2f.L.W.O/m=Fkg7bd,HcFEGb,IvlUe,MC8mtf,OF7gzc,RMhBfe,T4BAC,TJw5qb,TbaHGc,Y33vzc,cdos,hsm,iDPoPb,jsa,mvYTse,tg8oTe,uz938c,vWNDde,ws9Tlc,yQ43ff,d,csi/am=AAAAi4CdbgJB_r8CANgFACBwF9jIAhEIyFA1AAE/d=1/dg=2/br=1/rs=ACT90oHLt6EenBVAvKo6i3GbKovDrprVVw"}
       {:columnNumber 297,
        :functionName "s_df.addCallback",
        :lineNumber 373,
        :scriptId "25",
        :url
        "https://www.google.com/xjs/_/js/k=xjs.s.ru.zL4ese5rBGM.O/ck=xjs.s.-u89cqk529a2f.L.W.O/m=Fkg7bd,HcFEGb,IvlUe,MC8mtf,OF7gzc,RMhBfe,T4BAC,TJw5qb,TbaHGc,Y33vzc,cdos,hsm,iDPoPb,jsa,mvYTse,tg8oTe,uz938c,vWNDde,ws9Tlc,yQ43ff,d,csi/am=AAAAi4CdbgJB_r8CANgFACBwF9jIAhEIyFA1AAE/d=1/dg=2/br=1/rs=ACT90oHLt6EenBVAvKo6i3GbKovDrprVVw"}
       {:columnNumber 390,
        :functionName "s_2ra",
        :lineNumber 461,
        :scriptId "25",
        :url
        "https://www.google.com/xjs/_/js/k=xjs.s.ru.zL4ese5rBGM.O/ck=xjs.s.-u89cqk529a2f.L.W.O/m=Fkg7bd,HcFEGb,IvlUe,MC8mtf,OF7gzc,RMhBfe,T4BAC,TJw5qb,TbaHGc,Y33vzc,cdos,hsm,iDPoPb,jsa,mvYTse,tg8oTe,uz938c,vWNDde,ws9Tlc,yQ43ff,d,csi/am=AAAAi4CdbgJB_r8CANgFACBwF9jIAhEIyFA1AAE/d=1/dg=2/br=1/rs=ACT90oHLt6EenBVAvKo6i3GbKovDrprVVw"}
       {:columnNumber 32,
        :functionName "",
        :lineNumber 460,
        :scriptId "25",
        :url
        "https://www.google.com/xjs/_/js/k=xjs.s.ru.zL4ese5rBGM.O/ck=xjs.s.-u89cqk529a2f.L.W.O/m=Fkg7bd,HcFEGb,IvlUe,MC8mtf,OF7gzc,RMhBfe,T4BAC,TJw5qb,TbaHGc,Y33vzc,cdos,hsm,iDPoPb,jsa,mvYTse,tg8oTe,uz938c,vWNDde,ws9Tlc,yQ43ff,d,csi/am=AAAAi4CdbgJB_r8CANgFACBwF9jIAhEIyFA1AAE/d=1/dg=2/br=1/rs=ACT90oHLt6EenBVAvKo6i3GbKovDrprVVw"}
       {:columnNumber 328,
        :functionName "s_Vfa",
        :lineNumber 376,
        :scriptId "25",
        :url
        "https://www.google.com/xjs/_/js/k=xjs.s.ru.zL4ese5rBGM.O/ck=xjs.s.-u89cqk529a2f.L.W.O/m=Fkg7bd,HcFEGb,IvlUe,MC8mtf,OF7gzc,RMhBfe,T4BAC,TJw5qb,TbaHGc,Y33vzc,cdos,hsm,iDPoPb,jsa,mvYTse,tg8oTe,uz938c,vWNDde,ws9Tlc,yQ43ff,d,csi/am=AAAAi4CdbgJB_r8CANgFACBwF9jIAhEIyFA1AAE/d=1/dg=2/br=1/rs=ACT90oHLt6EenBVAvKo6i3GbKovDrprVVw"}
       {:columnNumber 238,
        :functionName "s_ff",
        :lineNumber 374,
        :scriptId "25",
        :url
        "https://www.google.com/xjs/_/js/k=xjs.s.ru.zL4ese5rBGM.O/ck=xjs.s.-u89cqk529a2f.L.W.O/m=Fkg7bd,HcFEGb,IvlUe,MC8mtf,OF7gzc,RMhBfe,T4BAC,TJw5qb,TbaHGc,Y33vzc,cdos,hsm,iDPoPb,jsa,mvYTse,tg8oTe,uz938c,vWNDde,ws9Tlc,yQ43ff,d,csi/am=AAAAi4CdbgJB_r8CANgFACBwF9jIAhEIyFA1AAE/d=1/dg=2/br=1/rs=ACT90oHLt6EenBVAvKo6i3GbKovDrprVVw"}
       {:columnNumber 297,
        :functionName "s_df.addCallback",
        :lineNumber 373,
        :scriptId "25",
        :url
        "https://www.google.com/xjs/_/js/k=xjs.s.ru.zL4ese5rBGM.O/ck=xjs.s.-u89cqk529a2f.L.W.O/m=Fkg7bd,HcFEGb,IvlUe,MC8mtf,OF7gzc,RMhBfe,T4BAC,TJw5qb,TbaHGc,Y33vzc,cdos,hsm,iDPoPb,jsa,mvYTse,tg8oTe,uz938c,vWNDde,ws9Tlc,yQ43ff,d,csi/am=AAAAi4CdbgJB_r8CANgFACBwF9jIAhEIyFA1AAE/d=1/dg=2/br=1/rs=ACT90oHLt6EenBVAvKo6i3GbKovDrprVVw"}
       {:columnNumber 2,
        :functionName "s_Zra",
        :lineNumber 460,
        :scriptId "25",
        :url
        "https://www.google.com/xjs/_/js/k=xjs.s.ru.zL4ese5rBGM.O/ck=xjs.s.-u89cqk529a2f.L.W.O/m=Fkg7bd,HcFEGb,IvlUe,MC8mtf,OF7gzc,RMhBfe,T4BAC,TJw5qb,TbaHGc,Y33vzc,cdos,hsm,iDPoPb,jsa,mvYTse,tg8oTe,uz938c,vWNDde,ws9Tlc,yQ43ff,d,csi/am=AAAAi4CdbgJB_r8CANgFACBwF9jIAhEIyFA1AAE/d=1/dg=2/br=1/rs=ACT90oHLt6EenBVAvKo6i3GbKovDrprVVw"}
       {:columnNumber 355,
        :functionName "s_1m.Ha",
        :lineNumber 458,
        :scriptId "25",
        :url
        "https://www.google.com/xjs/_/js/k=xjs.s.ru.zL4ese5rBGM.O/ck=xjs.s.-u89cqk529a2f.L.W.O/m=Fkg7bd,HcFEGb,IvlUe,MC8mtf,OF7gzc,RMhBfe,T4BAC,TJw5qb,TbaHGc,Y33vzc,cdos,hsm,iDPoPb,jsa,mvYTse,tg8oTe,uz938c,vWNDde,ws9Tlc,yQ43ff,d,csi/am=AAAAi4CdbgJB_r8CANgFACBwF9jIAhEIyFA1AAE/d=1/dg=2/br=1/rs=ACT90oHLt6EenBVAvKo6i3GbKovDrprVVw"}
       {:columnNumber 301,
        :functionName "s_lja.Da",
        :lineNumber 452,
        :scriptId "25",
        :url
        "https://www.google.com/xjs/_/js/k=xjs.s.ru.zL4ese5rBGM.O/ck=xjs.s.-u89cqk529a2f.L.W.O/m=Fkg7bd,HcFEGb,IvlUe,MC8mtf,OF7gzc,RMhBfe,T4BAC,TJw5qb,TbaHGc,Y33vzc,cdos,hsm,iDPoPb,jsa,mvYTse,tg8oTe,uz938c,vWNDde,ws9Tlc,yQ43ff,d,csi/am=AAAAi4CdbgJB_r8CANgFACBwF9jIAhEIyFA1AAE/d=1/dg=2/br=1/rs=ACT90oHLt6EenBVAvKo6i3GbKovDrprVVw"}
       {:columnNumber 343,
        :functionName "",
        :lineNumber 3,
        :scriptId "15",
        :url "https://www.google.com/?gws_rd=ssl"}]},
     :type "script"},
    :type "Other",
    :requestId "1000037133.32",
    :documentURL "https://www.google.com/?gws_rd=ssl",
    :loaderId "5790DDCB17D1ABDD2BEEA35F81CBEF31",
    :timestamp 1237876.471772}},
  :timestamp 1561989215810,
  :source nil,
  :datetime #inst "2019-07-01T13:53:35.810-00:00",
  :webview "71E8F527E41868DD41BCBF70AEFD7112",
  :_type :network/requestwillbesent}
 {:level ::info,
  :message
  {:method "Network.responseReceived",
   :params
   {:frameId "71E8F527E41868DD41BCBF70AEFD7112",
    :loaderId "5790DDCB17D1ABDD2BEEA35F81CBEF31",
    :requestId "1000037133.32",
    :response
    {:connectionId 331,
     :remotePort 443,
     :protocol "h2",
     :fromDiskCache false,
     :fromServiceWorker false,
     :connectionReused false,
     :headers
     {:date "Mon, 01 Jul 2019 13:42:16 GMT",
      :x-xss-protection "0",
      :server "gws",
      :content-length "0",
      :status "204",
      :x-frame-options "SAMEORIGIN",
      :strict-transport-security "max-age=31536000",
      :alt-svc "quic=\":443\"; ma=2592000; v=\"46,44,43,39\"",
      :content-type "text/html; charset=UTF-8"},
     :remoteIPAddress "74.125.131.105",
     :securityState "secure",
     :securityDetails
     {:keyExchangeGroup "X25519",
      :validFrom 1560256047,
      :protocol "TLS 1.3",
      :keyExchange "",
      :signedCertificateTimestampList
      [{:hashAlgorithm "SHA-256",
        :logDescription "Google 'Skydiver' log",
        :logId
        "BBD9DFBC1F8A71B593942397AA927B473857950AAB52E81A909664368E1ED185",
        :origin "TLS extension",
        :signatureAlgorithm "ECDSA",
        :signatureData
        "3044022074B0C0B4B627A4B19BDC80FB02AA58143DFD284AFD889EC8773E2F65AF55CE4F02200D97ED10102CEFA355B82254DF299ED1054AFDF16AE39E7BFC37BF836AECFFD1",
        :status "Verified",
        :timestamp 1.560258033442E12}
       {:hashAlgorithm "SHA-256",
        :logDescription "Sectigo 'Mammoth' CT log",
        :logId
        "6F5376AC31F03119D89900A45115FF77151C11D902C10029068DB2089A37D913",
        :origin "TLS extension",
        :signatureAlgorithm "ECDSA",
        :signatureData
        "30450220414D10B8BDC37BD9AE0C24336F2EF1F4B04A5BF8E4637833DF326CB2EB72DFE2022100D775929704C29A9CFE18E7D98AE37ADE83EFA800D2CAC17734F579537815B1CC",
        :status "Verified",
        :timestamp 1.560258033815E12}],
      :subjectName "www.google.com",
      :sanList ["www.google.com"],
      :certificateTransparencyCompliance "compliant",
      :validTo 1567513200,
      :issuer "Google Internet Authority G3",
      :certificateId 0,
      :cipher "AES_128_GCM"},
     :status 204,
     :timing
     {:workerReady -1,
      :pushStart 0,
      :workerStart -1,
      :sslEnd 59.846,
      :receiveHeadersEnd 87.467,
      :dnsStart 0.396,
      :sslStart 32.454,
      :requestTime 1237876.472484,
      :pushEnd 0,
      :proxyStart -1,
      :dnsEnd 8.557,
      :sendStart 60.284,
      :sendEnd 60.552,
      :proxyEnd -1,
      :connectEnd 59.856,
      :connectStart 8.557},
     :statusText "",
     :mimeType "text/html",
     :url
     "https://www.google.com/gen_204?atyp=csi&ei=gPkZXZSMG76HwPAPiPaw4AM&s=jsa&jsi=s,et.focus,n.iDPoPb,cn.5&zx=1561988536641",
     :encodedDataLength 163},
    :timestamp 1237876.560683,
    :type "Other"}},
  :timestamp 1561989215811,
  :source nil,
  :datetime #inst "2019-07-01T13:53:35.811-00:00",
  :webview "71E8F527E41868DD41BCBF70AEFD7112",
  :_type :network/responsereceived}
 {:level ::info,
  :message
  {:method "Network.loadingFinished",
   :params
   {:encodedDataLength 172,
    :requestId "1000037133.32",
    :shouldReportCorbBlocking false,
    :timestamp 1237876.560316}},
  :timestamp 1561989215811,
  :source nil,
  :datetime #inst "2019-07-01T13:53:35.811-00:00",
  :webview "71E8F527E41868DD41BCBF70AEFD7112",
  :_type :network/loadingfinished}]
```

**Comment by igrishaev** [comment]
OK @HeyItsMario try it now with the new functions mentioned here:
https://github.com/igrishaev/etaoin#devtools-tracking-http-requests-xhr-ajax

and bump the version to `0.3.5`
